app/utils/sql/SqlTypeImplicits.scala

```diff
@@ -1,35 +1,110 @@
+import play.api.libs.json.JsValue
+import slick.jdbc.{GetResult, PositionedResult}
 import utils.ObjectId
 
+import scala.concurrent.duration.FiniteDuration
+
 trait SqlTypeImplicits {
-  implicit protected object SetObjectId extends SetParameter[ObjectId] {
-    def apply(v: ObjectId, pp: PositionedParameters): Unit = pp.setString(v.id)
-  }
 
-  implicit protected object SetObjectIdOpt extends SetParameter[Option[ObjectId]] {
-    def apply(v: Option[ObjectId], pp: PositionedParameters): Unit = pp.setStringOption(v.map(_.id))
-  }
+  // reading typed results
 
   implicit protected object GetObjectId extends GetResult[ObjectId] {
     override def apply(v1: PositionedResult): ObjectId = ObjectId(v1.<<)
   }
 
-  implicit protected object SetInstant extends SetParameter[Instant] {
-    def apply(v: Instant, pp: PositionedParameters): Unit = pp.setTimestamp(v.toSql)
-  }
-
-  implicit protected object SetInstantOpt extends SetParameter[Option[Instant]] {
-    def apply(v: Option[Instant], pp: PositionedParameters): Unit = pp.setTimestampOption(v.map(_.toSql))
-  }
-
   implicit protected object GetInstant extends GetResult[Instant] {
     override def apply(v1: PositionedResult): Instant = Instant.fromSql(v1.<<)
   }
 
   implicit protected object GetInstantOpt extends GetResult[Option[Instant]] {
     override def apply(v1: PositionedResult): Option[Instant] = v1.nextTimestampOption().map(Instant.fromSql)
   }
+
+  // Conversions of values to SqlValue
+
+  implicit def stringToSqlValue(v: String): SqlValue = StringValue(v)
+
+  implicit def booleanToSqlValue(v: Boolean): SqlValue = BooleanValue(v)
+
+  implicit def intToSqlValue(v: Int): SqlValue = IntValue(v)
+
+  implicit def longToSqlValue(v: Long): SqlValue = LongValue(v)
+
+  implicit def floatToSqlValue(v: Float): SqlValue = FloatValue(v)
+
+  implicit def doubleToSqlValue(v: Double): SqlValue = DoubleValue(v)
+
+  implicit def enumToSqlValue(v: Enumeration#Value): SqlValue = EnumerationValue(v)
+
+  implicit def objectIdToSqlValue(v: ObjectId): SqlValue = ObjectIdValue(v)
+
+  implicit def instantToSqlValue(v: Instant): SqlValue = InstantValue(v)
+
+  implicit def jsValueToSqlValue(v: JsValue): SqlValue = JsonValue(v)
+
+  implicit def optionToSqlValue[T](v: Option[T])(implicit innerConversion: T => SqlValue): SqlValue =
+    v match {
+      case Some(inner) => innerConversion(inner)
+      case None        => NoneValue()
+    }
+
+  implicit def stringIterableToSqlValue(v: Iterable[String]): SqlValue = ArrayValue(v.toList)
+
+  implicit def boundingBoxToSqlValue(v: BoundingBox): SqlValue = BoundingBoxValue(v)
+
+  implicit def vec3IntToSqlValue(v: Vec3Int): SqlValue = Vector3Value(v.toVec3Double)
+
+  implicit def vec3DoubleToSqlValue(v: Vec3Double): SqlValue = Vector3Value(v)
+
+  implicit def finiteDurationToSqlValue(v: FiniteDuration): SqlValue = DurationValue(v)
+
+  // Conversions to SqlToken. This forwarded conversion is needed as implicits are not automatically chained.
```

Comment:
I was wondering, if we actually need the `*toSqlValue` functions? Maybe we could just pass around tokens.

Reply:
I am unsure about this. It would work with tokens, but in the tuple method, tokens that are more complex than single values don’t really make sense. With this two-step conversion, the signature of tuple can take List[SqlValue] and ensure nothing more complex is passed. But I agree that the additional conversions are also not great.

Reply:
ok, fine.